compiler/infra/ForwardList.hpp

```diff
@@ -0,0 +1,684 @@
+#ifndef FORWARD_LIST_HPP
```

Comment:
Is ReferenceWrapper an extension of std, or simply a wrapper around it?

I don't really have a preference for naming, I just want it consistent for these kinds of extensions.  If I had to pick, I'd go with lowercase like deque.hpp introduced.